src/coreComponents/fieldSpecification/FieldSpecificationManager.cpp

```diff
@@ -64,6 +64,142 @@ void FieldSpecificationManager::expandObjectCatalogs()
   }
 }
 
+void FieldSpecificationManager::validateBoundaryConditions( MeshLevel & mesh ) const
+{
+  // loop over all the FieldSpecification of the XML file
+  this->forSubGroups< FieldSpecificationBase >( [&] ( FieldSpecificationBase const & fs )
+  {
+    localIndex isFieldNameFound = 0;
+    // map from set name to a flag (1 if targetSet not empty, 0 otherwise)
+    map< string, localIndex > isTargetSetNonEmpty;
+    // map from set name to a flag (1 if targetSet has been created, 0 otherwise)
+    map< string, localIndex > isTargetSetCreated;
+
+    // Step 1: collect all the set names in a map (this is made necessary by the "apply" loop pattern
+
+    array1d< string > const & setNames = fs.getSetNames();
+    for( localIndex i = 0; i < setNames.size(); ++i )
+    {
+      isTargetSetNonEmpty[setNames[i]] = 0;
+      isTargetSetCreated[setNames[i]] = 0;
+    }
+
+    // Step 2: apply the boundary condition
+
+    fs.apply< dataRepository::Group >( mesh,
+                                       [&]( FieldSpecificationBase const &,
+                                            string const & setName,
+                                            SortedArrayView< localIndex const > const & targetSet,
+                                            Group & targetGroup,
+                                            string const fieldName )
+    {
+      dataRepository::InputFlags const flag = fs.getWrapper< string >( FieldSpecificationBase::viewKeyStruct::fieldNameString() ).getInputFlag();
+
+      // 2.a) If we enter this loop, we know that the set has been created
+      //      Fracture/fault sets are created later and the "apply" call silently ignores them
+      isTargetSetCreated.at( setName ) = 1;
+
+      // 2.b) If the fieldName is registered on this target, we record it
+      //      Unfortunately, we need two exceptions:
+      //       - FieldSpecification that do not target a field, like Aquifer, Traction, Equilibrium, etc. For these, the check is not
+      // necessary (the user cannot mess up)
+      //       - Face boundary conditions that target cell-based quantities, like the face BC of the flow solvers
```

Comment:
should we exclude blank `fieldName` from the check? Maybe we should have all the special cases as a derived `FieldSpecification` instead of using the `fieldName` to specify a different type of BC?

see comment in [CompositionalMultiphaseBase.cpp](https://github.com/GEOSX/GEOSX/pull/2291/files#diff-a3f511da5cf9cbca1b89b3f09ed486c4928bf6e7a4fe736bb6de427a6a6b98ba)

Reply:
I think it would be a little bit cleaner, but we would still need to pass a `fieldName` like for the `Traction` [here](https://github.com/GEOSX/GEOSX/blob/3301d288bd10fb6f551c7e1766e12bb25bfe0eff/src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp#L696-L699) or the `Aquifer` [here](https://github.com/GEOSX/GEOSX/blob/3301d288bd10fb6f551c7e1766e12bb25bfe0eff/src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp#L778-L781) for instance.

A refactoring to facilitate checks and clarify the behavior could be as follows:

1. We don't rely on `FieldSpecificationBase` anymore to apply the field boundary conditions, but we create a new type (maybe just `FieldSpecification` inherited from `FieldSpecificationBase` or something better) to allow us to distinguish the "real" field applications (say, pressure, temperature, displacement) from the "complex" boundary conditions (`Traction`, `Aquifer`, etc). Potentially, we can do that without changing the xml.

2. We would have a) a `FieldSpecificationManager::apply` for the "real" field applications that would take the `fieldName` or even the extrinsic data struct and b) FieldSpecificationManager::apply` for the "complex" field applications that would not require a field name.

Using Points 1 and 2, we could also strictly check the field name in the validation step because we would have a mechanism to know when the `fieldName` is actually a field name and not the catalog name of the boundary condition class.

Then if we want to go further:

3. We add a mechanism to identify the face boundary conditions that do not target a field registered on the `FaceManager`. For instance, we would have a class `FlowFaceDirichletBoundaryCondition` inherited from `FieldSpecificationBase`, and we would know that for this type, we should check the field name on the element regions. This requires changing the xml. This is just a nice to have, because we already have a lot of custom checks and logs for the face BCs in the flow solvers.

4. I would be nice to enforce that all the `FieldSpecification`s that have `initialCondition="1"` can actually be applied at initialization, which is not the case when we have the `SurfaceGenerator` (they either have an empty set, or a not existent set). A typical example is [here](https://github.com/GEOSX/GEOSX/blob/3301d288bd10fb6f551c7e1766e12bb25bfe0eff/inputFiles/singlePhaseFlowFractures/fractureMatrixFlow_edfm_base.xml#L97). This is the main source of difficulties and it also what complicates the validation of the initial/boundary conditions. 

Using all this points, an added bonus is that we can simplify the content of the `apply` functions. For instance, we could throw an error if the set does not exist [here](https://github.com/GEOSX/GEOSX/blob/3301d288bd10fb6f551c7e1766e12bb25bfe0eff/src/coreComponents/fieldSpecification/FieldSpecificationBase.hpp#L138), we would not have to check for the empty field name [here](https://github.com/GEOSX/GEOSX/blob/3301d288bd10fb6f551c7e1766e12bb25bfe0eff/src/coreComponents/fieldSpecification/FieldSpecificationManager.hpp#L219), and we could remove the check on the initial condition [here](https://github.com/GEOSX/GEOSX/blob/3301d288bd10fb6f551c7e1766e12bb25bfe0eff/src/coreComponents/fieldSpecification/FieldSpecificationManager.hpp#L220). So, less extra validation would be needed and the behavior of the functions would be more transparent.